### PR TITLE
Wrap `WorkerID` and `NodeID` / robust ID support

### PR DESCRIPTION
**Pull request by omus**
While working on expanding the exceptions we handle from the Raylet I found an [exception which required us to wrap `WorkerID`](https://github.com/ray-project/ray/blob/ray-2.5.1/python/ray/exceptions.py#L518). During implementing this I kept running into the same old issues such as handling `==` so I decided to refactor this code to be shared among all of he various ID types and also make the code robust.

### CODE REVIEW

File: src/ray_julia_jll/common.jl [hunk 1]
```diff
@@ -191,9 +190,21 @@ end
...
+    return JsonStringToMessage(Address, JSON3.write(nt))
+end
+
+function Base.show(io::IO, addr::Address)
+    raylet_hex = Hex(FromBinary(NodeID, raylet_id(addr)))
+    ip_addr_str = safe_convert(String, ip_address(addr))
+    worker_hex = Hex(FromBinary(WorkerID, worker_id(addr)))
+
+    print(io, "$Address((raylet_id=\"$raylet_hex\", ip_address=\"$ip_addr_str\", ")
+    print(io, "port=$(port(addr)), worker_id=\"$worker_hex\"))")
```
[the remark this answers is not in the document]
Comment (omus) [reply src/ray_julia_jll/common.jl:206]:
```suggestion
    print(io, "port=$(port(addr)), worker_id=\"$worker_hex\"))")
    return nothing
```

**Review by kleinschmidt — APPROVED** [review]

**Comment by kleinschmidt** [comment]
wanna bump the project.toml since we're doing releases now ;)

File: test/ray_julia_jll/id.jl [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        id = FromBinary(T, bytes)
+        @test length(bytes) == siz  # Bytes are not consumed by constructor
+        @test id isa T
+        @test Binary(Vector{UInt8}, id) == bytes
+        @test Binary(String, id) == bytes_str
+
+        id = FromBinary(T, UInt8[])
+        @test id isa T
+        @test Binary(Vector{UInt8}, id) == fill(0xff, siz)
+        @test Binary(String, id) == "\xff"^siz
```
[the remark this answers is not in the document]
Comment (omus) [reply test/ray_julia_jll/id.jl:71]: Note if you do this as a keyword the spaces are disallowed

File: src/ray_julia_jll/common.jl [hunk 3]
```diff
@@ -202,42 +214,101 @@ Base.show(io::IO, addr::Address) = print(io, _string(addr))
...
+
+        function FromHex(::Type{$T}, str::StdString)
+            if length(str) != 2 * Size($T)
+                msg = "Expected hex string length is 2 * $(Size($T)), provided length is $(length(str))"
+                throw(ArgumentError(msg))
+            elseif !ishex(str)
+                throw(ArgumentError("Expected hex string, found: $str"))
+            end
+            return $(Symbol(T, :FromHex))(str)
+        end
```
[the remark this answers is not in the document]
Comment (omus) [reply src/ray_julia_jll/common.jl:251]: I'll add a comment on this. The reasons for this are:

1. Better error handling as this exception has a type
2. The `FromBinary` exceptions from the C++ side will terminate the Julia process :S
3. In `FromHex` if you enter in a string of the right length but non-hex characters you'll end up being returned an ID with the hex value `"f"^Size(T)` which is awful

**Comment by omus** [comment]
> wanna bump the project.toml since we're doing releases now ;)

Will do.

File: test/ray_julia_jll/id.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (kleinschmidt) [reply test/ray_julia_jll/id.jl:71]: wat

[the remark this answers is not in the document]
Comment (omus) [reply test/ray_julia_jll/id.jl:71]: I was just mixed up. In all cases the JuliaFormatter disallows spaces around the `^` operator

**Pull request merged**